Multilanguage support

DanieleNeo opened the pull request:
Hi,

I immediately found this project very interesting, unfortunately for me it was in German, and to have a consultable interface I forked it and localized it in my language, unfortunately however in this way the synchronizations with the upstream risked being complicated, and potentially required important rework.

I therefore thought it could be useful to think of a multilingual implementation, based on a dictionary, also to facilitate those who are not native German speakers to approach the project.

To achieve this goal I acted on multiple levels, even if most of the interventions are on **__init__.py**.

I created a dictionary (currently three languages, German, English and Italian: I used German as a master and translated the others, using automatic tools and reserving for a later time, so as not to slow down the publication, the fine tuning of Italian one) and the related functions for setting the language and doing translations. Once set, during the building phase, only the chosen part of the dictionary is compiled as a resource for the cpp source together with the utility functions (translations.cpp file), this is necessary because it can be useful to have the dictionary also at runtime, but above all for the presence of some logic (controls in particular) that acted on hardcoded German strings, I did not want to upset the flow so translations were necessary. The dictionary on the runtime part does not seem to slow down the execution, in any case it is possible, through the python function, to limit only the dictionary records actually needed.

Furthermore, I also inserted a utility that if "auto" is set as the entity name in the yaml file, the same is automatically retrieved from the dictionary based on the language set.

`#mode
    ext:
      name: auto
#RT
#Pump
#EHS
#BPV
    tv:
      name: auto
      update_interval: 5`

It is possible to insert other languages, just act on the dictionary, if an entry for which translation is requested is not present for the language the fallback is English, and if it is not present here too an error message will be returned. It is good practice, when you insert new content or modify existing content, to also update the other languages, possibly with automatic tools. I am available to try to keep the Italian dictionary updated ;)

I hope I have done you a favor, as did the person who set up this project for me.

D.

wrfz commented:
> Hi,
> 
> I immediately found this project very interesting, unfortunately for me it was in German, and to have a consultable interface I forked it and localized it in my language, unfortunately however in this way the synchronizations with the upstream risked being complicated, and potentially required important rework.
> 
> I therefore thought it could be useful to think of a multilingual implementation, based on a dictionary, also to facilitate those who are not native German speakers to approach the project.
> 
> To achieve this goal I acted on multiple levels, even if most of the interventions are on **init.py**.
> 
> I created a dictionary (currently three languages, German, English and Italian: I used German as a master and translated the others, using automatic tools and reserving for a later time, so as not to slow down the publication, the fine tuning of Italian one) and the related functions for setting the language and doing translations. Once set, during the building phase, only the chosen part of the dictionary is compiled as a resource for the cpp source together with the utility functions (translations.cpp file), this is necessary because it can be useful to have the dictionary also at runtime, but above all for the presence of some logic (controls in particular) that acted on hardcoded German strings, I did not want to upset the flow so translations were necessary. The dictionary on the runtime part does not seem to slow down the execution, in any case it is possible, through the python function, to limit only the dictionary records actually needed.
> 
> Furthermore, I also inserted a utility that if "auto" is set as the entity name in the yaml file, the same is automatically retrieved from the dictionary based on the language set.
> 
> `#mode ext: name: auto #RT #Pump #EHS #BPV tv: name: auto update_interval: 5`
> 
> It is possible to insert other languages, just act on the dictionary, if an entry for which translation is requested is not present for the language the fallback is English, and if it is not present here too an error message will be returned. It is good practice, when you insert new content or modify existing content, to also update the other languages, possibly with automatic tools. I am available to try to keep the Italian dictionary updated ;)
> 
> I hope I have done you a favor, as did the person who set up this project for me.
> 
> D.

Great work!
One request: Please change the target branch from `main` to `dev`.
We always develop in the development branch and occasionally release to main.

DanieleNeo commented:
I done it, my mistake sorry.

DanieleNeo commented:
Hi @wrfz, wait for merging I'm experiencing a strange behavior with the optimized defrost, and I'm trying to understand if it's a regression (even if it seems unlikely considering the type of changes made) or my heat pump that is playing tricks on me: my device went into defrost twice, it went into summer mode for about 5 minutes but I didn't get a defrost notification and, above all, the BPV valve remained unchanged sending the heat from the accumulation to the radiant. I repeat, it's strange, it went into summer mode, I don't want it to be my HPSU that is playing a trick on me but I have to make sure. I changed the language to German and I'm testing this version, but unfortunately the outside temperature went up by 2 degrees and it hasn't been defrosting for a couple of hours. I'll continue the investigation tomorrow

wrfz commented:
I wasn't planning to merge the change yet, as long as you haven't addressed/responded to my comments.

What do you mean by "defrost notification"?
You can determine if the defrost optimization is functioning properly by observing the following values:

- **Betriebsart (mode of operating)** -> switches between **heating** and **defrosting**
- **Betriebsmodus (Operating mode)** -> switches between **heating** and **summer**
- **BPV** should switch to 100% during the defrost or summer phase (to ensure that the hot water from the tank is not pumped into the underfloor heating).

![grafik](https://github.com/user-attachments/assets/63538281-54dc-4c55-8c7f-597315031093)

DanieleNeo commented:
> I wasn't planning to merge the change yet, as long as you haven't addressed/responded to my comments.

Which comments? I have not seen any comments nor received any notifications, if you tell me where to look I will be happy to answer you

> 
> What do you mean by "defrost notification"? You can determine if the defrost optimization is functioning properly by observing the following values:
> 
> * **Betriebsart (mode of operating)** -> switches between **heating** and **defrosting**
> * **Betriebsmodus (Operating mode)** -> switches between **heating** and **summer**
> * **BPV** should switch to 100% during the defrost or summer phase (to ensure that the hot water from the tank is not pumped into the underfloor heating).
> 

So when I switched to the german builded version (multilanguage with de language specified) the optimized defrost worked properly and this was the situation

![image](https://github.com/user-attachments/assets/09b7f09d-c461-462b-854a-d0baecfa9819)

(the pink lines in Betriebsart is the defrost status, I don't know why it looks like this to me, maybe because I have the CAN and UART sensors on the same device with the same name)

this is the situation with the Italian version instead

![image](https://github.com/user-attachments/assets/79a63fea-9aac-4a2d-bd9c-9cade9d4c53f)

where:
- "Modalità di Operatività" is Betriebsart
- "Modalità di Funzionamento" is Betriebsmodus 
- "Valvola di ByPass" is BPV

I added also "Sbrinamento" of the UART control that stands for Abtauvorgang. As you can Mode of operating switch from Riscaldamento (heating) to Sbrinamento (defrosting), operating mode from Riscaldamento (Heating) to Estate (summer) but BPV remain to 0, so optimized defrosting is not working in this way.

For translations I start from German, create the keys in a concise way from English and proceed for the other languages ​​(currently only Italian), I think that in this phase there is a reduction of information that in the translation phase, when the name of the entities is set automatically, can give rise to conflicts or masking.

![image](https://github.com/user-attachments/assets/721b344b-3672-4120-96cd-d9354723a182)

as in this case, where select and sensor have the same name. It this is the problem can be easily solved with a fine tuning of translate dictionary, but it's just a hypothesis at the moment, I'm still investigating I thought it would be useful to share a partial analysis with you to also get your feedback.

D.

wrfz commented:
> Which comments? I have not seen any comments nor received any notifications, if you tell me where to look I will be happy to answer you

Please scroll up

DanieleNeo commented:
> > Which comments? I have not seen any comments nor received any notifications, if you tell me where to look I will be happy to answer you
> 
> Please scroll up

This?

![image](https://github.com/user-attachments/assets/0925613f-a3b9-4b04-84ee-8f1014196595)

I have done 3 days ago.

D.

DanieleNeo commented on components/daikin_rotex_can/daikin_rotex_can.cpp:
```diff
@@ -114,9 +115,9 @@ void DaikinRotexCanComponent::update_thermal_power() {
 
             float value = 0;
-            if (p_betriebs_art->state == "Warmwasserbereitung" && tv != nullptr && tr != nullptr) {
+            if (p_betriebs_art->state == translate("hot_water_production") && tv != nullptr && tr != nullptr) {
```
done

DanieleNeo commented on components/daikin_rotex_can/__init__.py:
```diff
@@ -11,9 +11,670 @@
+        _LOGGER.info("[Translate] Setting language to '%s'", lang)
+        current_language = lang
+    else:
+        _LOGGER.warning("[Translate] Language '%s' not found in dictionary. Falling back to English.", lang)
```
I solved it in a different way: I left the set_language and translate functions unchanged (in the future having an incomplete dictionary and a fallback on English could make sense), I defined a validation function (check_translations_integrity) that checks the integrity of the dictionary at build time and if not raises an exception with a report of the missing keys for each language

wrfz commented on components/daikin_rotex_can/translations.cpp:
```diff
@@ -0,0 +1,122 @@
+    {"off", "Spento"},
+    {"on", "Acceso"},
+    {"night_only", "Solo di Notte"},
+    {"monday", "Luned�"},
```
Seems the file is not saved as UTF-8. This could lead to issues. Please check this.

DanieleNeo replied:
this file should not have been pushed to github since the python init file dynamically creates it, I did it by mistake since it was an artifact of a test session build, in any case it served to highlight this situation and to move suggestion about map adoption. It was python that chose which format to adopt, now I forced it to always use UTF-8.

wrfz commented on components/daikin_rotex_can/daikin_rotex_can.cpp:
```diff
@@ -336,14 +343,14 @@ std::string DaikinRotexCanComponent::recalculate_state(EntityBase* pEntity, std:
                 ESP_LOGE(TAG, "3UV DHW defekt => tvbh: %f, tv: %f, max_spread: %f, bpv: %f, flow_rate: %f",
                     tvbh->state, tv->state, m_max_spread.tvbh_tv, dhw_mixer_position->state, flow_rate->state);
-                return new_state + "|3UV DHW defekt";
+                return new_state + "|3UV DHW " + translate("defect");
```
Could be also a constant

DanieleNeo replied:
I had missed it, now it's a constant

wrfz commented on components/daikin_rotex_can/daikin_rotex_can.cpp:
```diff
@@ -336,14 +343,14 @@ std::string DaikinRotexCanComponent::recalculate_state(EntityBase* pEntity, std:
                 ESP_LOGE(TAG, "3UV BPV defekt => tvbh: %f, tr: %f, max_spread: %f, dhw_mixer_pos: %f, flow_rate: %f",
                     tvbh->state, tr->state, m_max_spread.tvbh_tr, bpv->state, flow_rate->state);
-                return new_state + "|3UV BPV defekt";
+                return new_state + "|3UV BPV "+ translate("defect");
```
constant?

wrfz commented on components/daikin_rotex_can/translations.cpp:
```diff
@@ -0,0 +1,122 @@
+namespace esphome {
+namespace daikin_rotex_can {
+
+static const Translation translations[] = {
```
Could be a map?

DanieleNeo replied:
I had adopted a simple structure to avoid having problems with memory usage when the dictionary was larger, but in any case I had not thought of map, which I have now implemented by adapting the translate function

wrfz commented on examples/minimal.yaml:
```diff
@@ -43,4 +43,4 @@ daikin_rotex_can:
   entities:
     tdhw1:
-      name: Warmwassertemperatur
+      name: Warmwassertemperatur
```
keep new line at the end?

wrfz commented on examples/options.yaml:
```diff
@@ -21,4 +21,4 @@ daikin_rotex_can:
         20: 20 °C
         21: 21 °C
-        22: 22 °C
+        22: 22 °C
```
new line?

wrfz replied:
Please check all files

wrfz commented on examples/full_en.yaml:
```diff
@@ -0,0 +1,326 @@
+    tdhw1:
+      name: Hot Water Temperature
+    target_supply_temperature:
+      name: Target Supply Temperature
```
new line?

wrfz commented on components/daikin_rotex_can/translations.cpp:
```diff
@@ -0,0 +1,122 @@
+    {"sg3_hot_water_70c", "SG3 - Acqua Calda 70�C"}
+};
+
+std::string translate(const std::string &key) {
```
If translations would be a Map, then there would be no need to search.
Also better access performance => O(log n) vs O(n)
https://www.bigocheatsheet.com/

DanieleNeo replied:
done, as said above

wrfz commented on components/daikin_rotex_can/translations/translate.py:
```diff
@@ -0,0 +1,153 @@
+    
+    _LOGGER.info("[Translate] All translations are consistent!")
+
+def set_language(lang):
```
dead code?

wrfz commented on components/daikin_rotex_can/__init__.py:
```diff
@@ -51,8 +63,8 @@
         "map": {
-            0x00: "Aus",
-            0x01: "An"
+            0x00: delayed_translate("off"),
```
Such recurring definitions could also be constants here.

DanieleNeo replied:
It doesn't make much sense in python, ok for cpp where resources are limited and optimization is important, but the python script is run once in configuration, maybe it's easier to copy&paste it that way

wrfz commented on components/daikin_rotex_can/translations/translate.py:
```diff
@@ -0,0 +1,153 @@
+    global current_language
+    _LOGGER.info("Writing cpp translate file")
+    cpp_code = generate_cpp_translations_for_language(translations, current_language)
+    output_path = os.path.join(output_dir, "translations.cpp")
```
This file should be added to .gitignore or generated girectly in build directory.
Otherwise it appears as an untracked file in git.

DanieleNeo replied:
done (.gitignore)

wrfz commented on components/daikin_rotex_can/translations/translate.py:
```diff
@@ -0,0 +1,153 @@
+def set_language(lang):
+    global current_language
+    if lang in translations:
+        _LOGGER.info("[Translate] Setting language to '%s'", lang)
```
This output list is too intrusive/extensive. It was helpful during translation development. Later on, it will get in the way because the other information will be lost. It is also relatively uninteresting for users.
Suggest to remove the _LOGGER.info output here.

DanieleNeo replied:
downgrade info to debug level (for translation function, set_language is one occurency and could be useful to check if set language was taken)

DanieleNeo commented:
> Tested DE variant. It crashes directly at startup.
> 
> ![grafik](https://private-user-images.githubusercontent.com/77174685/401666748-225b2d32-5efd-4160-9379-49ab83ac920b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0NTExNTksIm5iZiI6MTczNjQ1MDg1OSwicGF0aCI6Ii83NzE3NDY4NS80MDE2NjY3NDgtMjI1YjJkMzItNWVmZC00MTYwLTkzNzktNDlhYjgzYWM5MjBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDE5MjczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4YjkxYzU0ZDBkM2EwNmQyOGZiMTEyOTlhODlkMzQ0NWI4NjZkMDk0MGI4N2U0ZjlhMzY4ZTdhZTZkOGQ0ZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XxU1HXg4EQLsuFDgEd2H1Sbpj0WJZAaSgXuINsR6xsE)

Are you sure you're using latest source, without any typos, like the translate.cpp file I accidentally included in github? I also tried with De variant, this is my log

> INFO ESPHome 2024.12.2
INFO Reading configuration /config/esphome/hpsurotex.yaml...
INFO Updating https://github.com/DanieleNeo/Daikin-Rotex-HPSU-CAN-Seriell.git@feature/localization
INFO Updating https://github.com/DanieleNeo/esphome-components@feature/localization
INFO [Translate] All translations are consistent!
INFO [Translate] All translations are consistent!
INFO Project Git Hash 1a2989b
INFO Detected timezone 'Europe/Rome'
INFO Generating C++ source...
INFO [Translate] Setting language to 'de'
INFO [Translate] Key 'standby' found in language 'de' -> 'Standby'
INFO [Translate] Key 'heating' found in language 'de' -> 'Heizen'
INFO [Translate] Key 'cooling' found in language 'de' -> 'Kühlen'
INFO [Translate] Key 'hot_water' found in language 'de' -> 'Warmwasserbereitung'
INFO [Translate] Key 'heating_hot_water' found in language 'de' -> 'Heizen + Warmwasser'
INFO [Translate] Key 'cooling_hot_water' found in language 'de' -> 'Kühlen + Warmwasser'
INFO [Translate] Setting language to 'de'
INFO Writing cpp translate file
INFO Building cpp translate dictionary for language: de
INFO Generated /data/external_components/6908031a/components/daikin_rotex_can/translations.cpp
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'on' found in language 'de' -> 'An'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'monday' found in language 'de' -> 'Montag'
INFO [Translate] Key 'tuesday' found in language 'de' -> 'Dienstag'
INFO [Translate] Key 'wednesday' found in language 'de' -> 'Mittwoch'
INFO [Translate] Key 'thursday' found in language 'de' -> 'Donnerstag'
INFO [Translate] Key 'friday' found in language 'de' -> 'Freitag'
INFO [Translate] Key 'saturday' found in language 'de' -> 'Samstag'
INFO [Translate] Key 'sunday' found in language 'de' -> 'Sonntag'
INFO [Translate] Key 'mo_to_su' found in language 'de' -> 'Montag bis Sonntag'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'on' found in language 'de' -> 'An'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'standby' found in language 'de' -> 'Standby'
INFO [Translate] Key 'heating' found in language 'de' -> 'Heizen'
INFO [Translate] Key 'cooling' found in language 'de' -> 'Kühlen'
INFO [Translate] Key 'defrosting' found in language 'de' -> 'Abtauen'
INFO [Translate] Key 'hot_water_production' found in language 'de' -> 'Warmwasserbereitung'
INFO [Translate] Key 'standby' found in language 'de' -> 'Standby'
INFO [Translate] Key 'heating' found in language 'de' -> 'Heizen'
INFO [Translate] Key 'lowering' found in language 'de' -> 'Absenken'
INFO [Translate] Key 'summer' found in language 'de' -> 'Sommer'
INFO [Translate] Key 'cooling' found in language 'de' -> 'Kühlen'
INFO [Translate] Key 'automatic_1' found in language 'de' -> 'Automatik 1'
INFO [Translate] Key 'automatic_2' found in language 'de' -> 'Automatik 2'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'on' found in language 'de' -> 'An'
INFO [Translate] Key 'night_only' found in language 'de' -> 'Nur bei Nacht'
INFO [Translate] Key 'err_0' found in language 'de' -> 'Kein Fehler'
INFO [Translate] Key 'err_E9001' found in language 'de' -> 'E9001 Rücklauffühler'
INFO [Translate] Key 'err_E9002' found in language 'de' -> 'E9002 Vorlauffühler'
INFO [Translate] Key 'err_E9003' found in language 'de' -> 'E9003 Frostschutzfunktion'
INFO [Translate] Key 'err_E9004' found in language 'de' -> 'E9004 Durchfluss'
INFO [Translate] Key 'err_E9005' found in language 'de' -> 'E9005 Vorlauftemperaturfühler'
INFO [Translate] Key 'err_E9006' found in language 'de' -> 'E9006 Vorlauftemperaturfühler'
INFO [Translate] Key 'err_E9007' found in language 'de' -> 'E9007 Platine IG defekt'
INFO [Translate] Key 'err_E9008' found in language 'de' -> 'E9008 Kältemitteltemperatur außerhalb des Bereiches'
INFO [Translate] Key 'err_E9009' found in language 'de' -> 'E9009 STB Fehler'
INFO [Translate] Key 'err_E9010' found in language 'de' -> 'E9010 STB Fehler'
INFO [Translate] Key 'err_E9011' found in language 'de' -> 'E9011 Fehler Flowsensor'
INFO [Translate] Key 'err_E9012' found in language 'de' -> 'E9012 Fehler Vorlauffühler'
INFO [Translate] Key 'err_E9013' found in language 'de' -> 'E9013 Platine AG defekt'
INFO [Translate] Key 'err_E9014' found in language 'de' -> 'E9014 P-Kältemittel hoch'
INFO [Translate] Key 'err_E9015' found in language 'de' -> 'E9015 P-Kältemittel niedrig'
INFO [Translate] Key 'err_E9016' found in language 'de' -> 'E9016 Lastschutz Verdichter'
INFO [Translate] Key 'err_E9017' found in language 'de' -> 'E9017 Ventilator blockiert'
INFO [Translate] Key 'err_E9018' found in language 'de' -> 'E9018 Expansionsventil'
INFO [Translate] Key 'err_E9019' found in language 'de' -> 'E9019 Warmwassertemperatur > 85°C'
INFO [Translate] Key 'err_E9020' found in language 'de' -> 'E9020 T-Verdampfer hoch'
INFO [Translate] Key 'err_E9021' found in language 'de' -> 'E9021 HPS-System'
INFO [Translate] Key 'err_E9022' found in language 'de' -> 'E9022 Fehler AT-Fühler'
INFO [Translate] Key 'err_E9023' found in language 'de' -> 'E9023 Fehler WW-Fühler'
INFO [Translate] Key 'err_E9024' found in language 'de' -> 'E9024 Drucksensor'
INFO [Translate] Key 'err_E9025' found in language 'de' -> 'E9025 Fehler Rücklauffühler'
INFO [Translate] Key 'err_E9026' found in language 'de' -> 'E9026 Drucksensor'
INFO [Translate] Key 'err_E9027' found in language 'de' -> 'E9027 Aircoil-Fühler Defrost'
INFO [Translate] Key 'err_E9028' found in language 'de' -> 'E9028 Aircoil-Fühler temp'
INFO [Translate] Key 'err_E9029' found in language 'de' -> 'E9029 Fehler Kältefühler AG'
INFO [Translate] Key 'err_E9030' found in language 'de' -> 'E9030 Defekt elektrisch'
INFO [Translate] Key 'err_E9031' found in language 'de' -> 'E9031 Defekt elektrisch'
INFO [Translate] Key 'err_E9032' found in language 'de' -> 'E9032 Defekt elektrisch'
INFO [Translate] Key 'err_E9033' found in language 'de' -> 'E9033 Defekt elektrisch'
INFO [Translate] Key 'err_E9034' found in language 'de' -> 'E9034 Defekt elektrisch'
INFO [Translate] Key 'err_E9035' found in language 'de' -> 'E9035 Platine AG defekt'
INFO [Translate] Key 'err_E9036' found in language 'de' -> 'E9036 Defekt elektrisch'
INFO [Translate] Key 'err_E9037' found in language 'de' -> 'E9037 Einstellung Leistung'
INFO [Translate] Key 'err_E9038' found in language 'de' -> 'E9038 Kältemittel Leck'
INFO [Translate] Key 'err_E9039' found in language 'de' -> 'E9039 Unter/Überspannung'
INFO [Translate] Key 'err_E9041' found in language 'de' -> 'E9041 Übertragungsfehler'
INFO [Translate] Key 'err_E9042' found in language 'de' -> 'E9042 Übertragungsfehler'
INFO [Translate] Key 'err_E9043' found in language 'de' -> 'E9043 Übertragungsfehler'
INFO [Translate] Key 'err_E9044' found in language 'de' -> 'E9044 Übertragungsfehler'
INFO [Translate] Key 'err_E75' found in language 'de' -> 'E75 Fehler Außentemperaturfühler'
INFO [Translate] Key 'err_E76' found in language 'de' -> 'E76 Fehler Speichertemperaturfühler'
INFO [Translate] Key 'err_E81' found in language 'de' -> 'E81 Kommunikationsfehler Rocon'
INFO [Translate] Key 'err_E88' found in language 'de' -> 'E88 Kommunikationsfehler Rocon Handbuch'
INFO [Translate] Key 'err_E91' found in language 'de' -> 'E91 Kommunikationsfehler Rocon Handbuch'
INFO [Translate] Key 'err_E128' found in language 'de' -> 'E128 Fehler Rücklauftemperaturfühler'
INFO [Translate] Key 'err_E129' found in language 'de' -> 'E129 Fehler Drucksensor'
INFO [Translate] Key 'err_E198' found in language 'de' -> 'E198 Durchflussmessung nicht plausibel'
INFO [Translate] Key 'err_E200' found in language 'de' -> 'E200 Kommunikationsfehler'
INFO [Translate] Key 'err_E8005' found in language 'de' -> 'E8005 Wasserdruck in Heizungsanlage zu gering'
INFO [Translate] Key 'err_E8100' found in language 'de' -> 'E8100 Kommunikation'
INFO [Translate] Key 'err_E9000' found in language 'de' -> 'E9000 Interne vorübergehende Meldung'
INFO [Translate] Key 'err_W8006' found in language 'de' -> 'W8006 Warnung Druckverlust'
INFO [Translate] Key 'err_W8007' found in language 'de' -> 'W8007 Wasserdruck in Anlage zu hoch'
INFO [Translate] Key 'weather_dependent' found in language 'de' -> 'Witterungsgeführt'
INFO [Translate] Key 'fixed' found in language 'de' -> 'Fest'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'sg_mode_1' found in language 'de' -> 'SG Modus 1'
INFO [Translate] Key 'sg_mode_2' found in language 'de' -> 'SG Modus 2'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'on' found in language 'de' -> 'An'
INFO [Translate] Key 'no_additional_heat_generator' found in language 'de' -> 'Kein zusätzlicher Wärmeerzeuger'
INFO [Translate] Key 'optional_backup_heater' found in language 'de' -> 'Optionaler Backup-Heater'
INFO [Translate] Key 'wez_for_hot_water_and_heating' found in language 'de' -> 'WEZ für WW und HZ'
INFO [Translate] Key 'wez1_for_hot_water_wez2_for_heating' found in language 'de' -> 'WEZ1 für WW - WEZ2 für HZ'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'on' found in language 'de' -> 'An'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'sgn_normal_mode' found in language 'de' -> 'SGN - Normaler Modus'
INFO [Translate] Key 'sg1_hot_water_and_heating_off' found in language 'de' -> 'SG1 - WW & HZ ausgeschalten'
INFO [Translate] Key 'sg2_hot_water_and_heating_plus_5c' found in language 'de' -> 'SG2 - WW & HZ + 5°C'
INFO [Translate] Key 'sg3_hot_water_70c' found in language 'de' -> 'SG3 - WW 70°C'
INFO [Translate] Key 'off' found in language 'de' -> 'Aus'
INFO [Translate] Key 'on' found in language 'de' -> 'An'
INFO Compiling app...
Processing hpsurotex (board: esp32-s3-devkitc-1; framework: arduino; platform: platformio/espressif32@5.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32S3 240MHz, 320KB RAM, 8MB Flash
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch5 
 - toolchain-xtensa-esp32s3 @ 8.4.0+2021r2-patch5
Dependency Graph
|-- AsyncTCP-esphome @ 2.1.4
|-- WiFi @ 2.0.0
|-- FS @ 2.0.0
|-- Update @ 2.0.0
|-- ESPAsyncWebServer-esphome @ 3.2.2
|-- DNSServer @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- noise-c @ 0.1.6
|-- ArduinoJson @ 6.18.5
Compiling .pioenvs/hpsurotex/src/esphome/components/daikin_rotex_can/translations.cpp.o
Compiling .pioenvs/hpsurotex/src/main.cpp.o
Linking .pioenvs/hpsurotex/firmware.elf
RAM:   [=         ]  14.3% (used 46748 bytes from 327680 bytes)
Flash: [=======   ]  72.5% (used 1330241 bytes from 1835008 bytes)
Building .pioenvs/hpsurotex/firmware.bin
Creating esp32s3 image...
Successfully created esp32s3 image.
esp32_create_combined_bin([".pioenvs/hpsurotex/firmware.bin"], [".pioenvs/hpsurotex/firmware.elf"])
Wrote 0x154dc0 bytes to file /data/build/hpsurotex/.pioenvs/hpsurotex/firmware.factory.bin, ready to flash to offset 0x0
esp32_copy_ota_bin([".pioenvs/hpsurotex/firmware.bin"], [".pioenvs/hpsurotex/firmware.elf"])
======================== [SUCCESS] Took 123.30 seconds ========================
INFO Successfully compiled program.
INFO Connecting to 10.0.22.139 port 3232...
INFO Connected to 10.0.22.139
INFO Uploading /data/build/hpsurotex/.pioenvs/hpsurotex/firmware.bin (1330624 bytes)
Uploading: [============================================================] 100% Done...

INFO Upload took 9.68 seconds, waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from 10.0.22.139 using esphome API
INFO Successfully connected to hpsurotex @ 10.0.22.139 in 7.241s
INFO Successful handshake with hpsurotex @ 10.0.22.139 in 0.066s
[20:32:33][I][app:100]: ESPHome version 2024.12.2 compiled on Jan  9 2025, 20:30:26
[20:32:33][C][wifi:600]: WiFi:
[20:32:33][C][wifi:428]:   Local MAC: 64:E8:33:7D:E9:9C
[20:32:33][C][wifi:433]:   SSID: 'BuffaWiFiHomeDomo'[redacted]
[20:32:33][C][wifi:436]:   IP Address: 10.0.22.139
[20:32:34][C][wifi:440]:   BSSID: 32:B5:C2:3E:7A:D8[redacted]
[20:32:34][C][wifi:441]:   Hostname: 'hpsurotex'
[20:32:34][C][wifi:443]:   Signal strength: -89 dB ▂▄▆█
[20:32:34][C][wifi:447]:   Channel: 1
[20:32:34][C][wifi:448]:   Subnet: 255.255.255.0
[20:32:34][C][wifi:449]:   Gateway: 10.0.22.1
[20:32:34][C][wifi:450]:   DNS1: 10.0.22.1
[20:32:34][C][wifi:451]:   DNS2: 0.0.0.0
[20:32:34][C][logger:185]: Logger:
[20:32:34][C][logger:186]:   Level: DEBUG
[20:32:34][C][logger:188]:   Log Baud Rate: 115200
[20:32:34][C][logger:189]:   Hardware UART: USB_CDC
[20:32:34][C][uart.arduino_esp32:151]: UART Bus 0:
[20:32:34][C][uart.arduino_esp32:152]:   TX Pin: GPIO17
[20:32:34][C][uart.arduino_esp32:153]:   RX Pin: GPIO16
[20:32:34][C][uart.arduino_esp32:155]:   RX Buffer Size: 256
[20:32:34][C][uart.arduino_esp32:160]:   Stop bits: 1
[20:32:34][C][uptime.sensor:033]: Uptime Sensor 'Uptime Seconds'
[20:32:34][C][uptime.sensor:032]: Uptime Sensor 'Uptime Timestamp'
[20:32:34][C][uptime.sensor:032]:   Device Class: 'timestamp'
[20:32:34][C][homeassistant.time:010]: Home Assistant Time:
[20:32:34][C][homeassistant.time:011]:   Timezone: 'CET-1CEST,M3.5.0,M10.5.0/3'
[20:32:34][C][restart.button:017]: Restart Button 'Rotex ESP Restart'
[20:32:34][C][restart.button:017]:   Icon: 'mdi:restart'
[20:32:34][C][daikin_rotex_can:306]: DaikinRotexCanComponent
[20:32:34][C][captive_portal:089]: Captive Portal:
[20:32:34][C][web_server:153]: Web Server:
[20:32:34][C][web_server:154]:   Address: hpsurotex.local:80
[20:32:34][C][mdns:116]: mDNS:
[20:32:34][C][mdns:117]:   Hostname: hpsurotex
[20:32:34][C][esphome.ota:073]: Over-The-Air updates:
[20:32:34][C][esphome.ota:074]:   Address: hpsurotex.local:3232
[20:32:34][C][esphome.ota:075]:   Version: 2
[20:32:34][C][esphome.ota:078]:   Password configured
[20:32:34][C][safe_mode:018]: Safe Mode:
[20:32:34][C][safe_mode:020]:   Boot considered successful after 60 seconds
[20:32:34][C][safe_mode:021]:   Invoke after 10 boot attempts
[20:32:34][C][safe_mode:023]:   Remain in safe mode for 300 seconds
[20:32:34][C][api:140]: API Server:
[20:32:34][C][api:141]:   Address: hpsurotex.local:6053
[20:32:34][C][api:143]:   Using noise encryption: YES
[20:32:34][C][wifi_signal.sensor:010]: WiFi Signal 'WiFi Signal Sensor'
[20:32:34][C][wifi_signal.sensor:010]:   Device Class: 'signal_strength'
[20:32:34][C][wifi_signal.sensor:010]:   State Class: 'measurement'
[20:32:34][C][wifi_signal.sensor:010]:   Unit of Measurement: 'dBm'
[20:32:34][C][wifi_signal.sensor:010]:   Accuracy Decimals: 0
[20:32:34][C][wifi_info:012]: WifiInfo SSID 'ESP Connected SSID'
[20:32:34][C][wifi_info:010]: WifiInfo IPAddress 'ESP IP Address'
[20:32:34][C][debug:021]: Debug component:
[20:32:34][C][debug:023]:   Device info 'Device Info'
[20:32:34][C][debug:023]:     Icon: 'mdi:chip'
[20:32:34][C][debug:026]:   Free space on heap 'Heap Free'
[20:32:34][C][debug:026]:     State Class: ''
[20:32:34][C][debug:026]:     Unit of Measurement: 'B'
[20:32:34][I][daikin-uart:133]: 9317| 0x10:-|5006|5034| 0x20:-|5040|5076| 0x21:-|5082|5101| 0x30:-|5107|5138| 0x60:-|5144|5173| 0x61:-|5179|5212
[20:32:35][D][sensor:094]: 'Potenza Termica Regolare': Sending state 0.42000 kW with 2 decimals of accuracy
[20:32:35][I][daikin-uart:133]: 10039| 0x10:R|5006|5034| 0x20:-|5040|5076| 0x21:-|5082|5101| 0x30:-|5107|5138| 0x60:-|5144|5173| 0x61:-|5179|5212
[20:32:35][I][daikin-uart:029]: TX: 03 40 10 AC
[20:32:35][D][sensor:094]: 'Temperatura di condensazione target': Sending state 35.10000 °C with 1 decimals of accuracy
[20:32:35][I][daikin-uart:077]: RX: 40 10 12 01 00 00 00 00 00 84 03 00 00 00 00 00 00 00 00 15|press_eq: |defrost_operation: |tgt_cond_temp: 35.1
[20:32:35][I][daikin-uart:133]: 10092| 0x10:-|10048|10088| 0x20:R|5040|5076| 0x21:-|5082|5101| 0x30:-|5107|5138| 0x60:-|5144|5173| 0x61:-|5179|5212
[20:32:35][I][daikin-uart:029]: TX: 03 40 20 9C
[20:32:35][D][sensor:094]: 'Temperatura aria esterna': Sending state 13.50000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura dopo compressore': Sending state 24.50000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura scambiatore lamellare': Sending state 12.50000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Pressione refrigerante': Sending state 10.70000  with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura refrigerante': Sending state 9.40000 °C with 1 decimals of accuracy
[20:32:35][I][daikin-uart:077]: RX: 40 20 13 87 00 00 00 F5 00 00 00 7D 00 00 00 BE 00 6B 00 00 6A|outdoor_temp: 13.5|temp_after_compr: 24.5|exch_temp: 12.5|liq_pressure: 10.7|t_liq2: 9.4
[20:32:35][I][daikin-uart:133]: 10155| 0x10:-|10048|10088| 0x20:-|10101|10152| 0x21:R|5082|5101| 0x30:R|5107|5138| 0x60:-|5144|5173| 0x61:-|5179|5212
[20:32:35][I][daikin-uart:029]: TX: 03 40 21 9B
[20:32:35][D][sensor:094]: 'Corrente primaria INV': Sending state 0.50000 A with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Corrente secondaria INV': Sending state 0.00000 A with 1 decimals of accuracy
[20:32:35][I][daikin-uart:077]: RX: 40 21 07 05 00 00 00 00 92|inv_prim_current: 0.5|inv_sec_current: 0.0
[20:32:35][I][daikin-uart:133]: 10189| 0x10:-|10048|10088| 0x20:-|10101|10152| 0x21:-|10164|10187| 0x30:R|5107|5138| 0x60:R|5144|5173| 0x61:-|5179|5212
[20:32:35][I][daikin-uart:029]: TX: 03 40 30 8C
[20:32:35][D][sensor:094]: 'Frequenza compressore': Sending state 0.00000 RPM with 0 decimals of accuracy
[20:32:35][D][sensor:094]: 'Velocità ventola': Sending state 0.00000 RPM with 0 decimals of accuracy
[20:32:35][D][sensor:094]: 'Valvola di espansione': Sending state 90.00000 % with 0 decimals of accuracy
[20:32:35][I][daikin-uart:077]: RX: 40 30 0D 00 00 00 C2 01 00 00 80 00 00 00 3F|inv_freq: 0|fan_speed: 0|exv: 90|4_way_valve: 
[20:32:35][I][daikin-uart:133]: 10250| 0x10:-|10048|10088| 0x20:-|10101|10152| 0x21:-|10164|10187| 0x30:-|10198|10246| 0x60:R|5144|5173| 0x61:R|5179|5212
[20:32:35][I][daikin-uart:029]: TX: 03 40 60 5C
[20:32:35][D][text_sensor:064]: 'Stato di funzionamento': Sending state 'Heizen'
[20:32:35][I][daikin-uart:077]: RX: 40 60 12 80 00 10 00 40 3E 4B EA 01 F0 00 A0 80 05 03 00 F1|mode_of_operating: Heizen|buh1: |buh2: |buh_bsh: 
[20:32:35][I][daikin-uart:133]: 10302| 0x10:-|10048|10088| 0x20:-|10101|10152| 0x21:-|10164|10187| 0x30:-|10198|10246| 0x60:-|10259|10299| 0x61:R|5179|5212
[20:32:35][I][daikin-uart:029]: TX: 03 40 61 5B
[20:32:35][I][daikin-uart:133]: 10328| 0x10:-|10048|10088| 0x20:-|10101|10152| 0x21:-|10164|10187| 0x30:-|10198|10246| 0x60:-|10259|10299| 0x61:P|10311|5212
[20:32:35][D][sensor:094]: 'Temperatura di mandata (TV)': Sending state 20.00000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura di mandata (TVBH)': Sending state 19.90000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura linea liquido': Sending state 20.80000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura di ritorno': Sending state 19.70000 °C with 1 decimals of accuracy
[20:32:35][D][sensor:094]: 'Temperatura serbatoio ACS': Sending state 47.10000 °C with 1 decimals of accuracy
[20:32:35][I][daikin-uart:077]: RX: 40 61 10 80 00 C8 00 C7 00 D0 00 C5 00 D7 01 00 00 D2|tv: 20.0|tv_bh: 19.9|t_liq: 20.8|tr: 19.7|dhw_temp: 47.1
[20:32:35][D][sensor:094]: 'Heap Free': Sending state 142360.00000 B with 0 decimals of accuracy
[20:32:35][D][sensor:094]: 'Heap Max Block': Sending state 129012.00000 B with 0 decimals of accuracy
[20:32:36][D][canbus:072]: received can message (#1) std can_id=0x180 size=7
[20:32:36][I][unhandled:146]: can_id<0x180> data<20 0A 52 00 00 00 00>
[20:32:36][I][daikin-uart:133]: 11340| 0x10:-|10048|10088| 0x20:-|10101|10152| 0x21:-|10164|10187| 0x30:-|10198|10246| 0x60:-|10259|10299| 0x61:-|10311|10364
[20:32:37][D][canbus:035]: send standard id=0x680 rtr=FALSE size=7
[20:32:37][I][sendGet:146]: Temperatura di mandata (TV) can_id<0x680> command<31 00 FA C0 FC 00 00>
[20:32:37][D][canbus:072]: received can message (#1) std can_id=0x180 size=7
[20:32:37][D][sensor:094]: 'Temperatura di mandata (TV)': Sending state 20.30000 °C with 1 decimals of accuracy
[20:32:37][I][handle :146]: Temperatura di mandata (TV)<20.299999> can_id<0x180> data<D2 00 FA C0 FC 00 CB> changed<0>
[20:32:37][D][canbus:035]: send standard id=0x680 rtr=FALSE size=7
[20:32:37][I][sendGet:146]: Temperatura Mandata Riscaldamento (TVBH) can_id<0x680> command<31 00 FA C0 FE 00 00>
[20:32:37][D][canbus:072]: received can message (#1) std can_id=0x180 size=7

wrfz commented:
Ganz sicher. Hatte den Commits mit dem github verglichen. War identisch.
Ich hatte es auch mehrmals hin und her probiert.
`dev` hat immer funktioniert und `feature/localization` crashte jedes mal.

Ich werde morgen versuchen das Problem zu analysieren.

DanieleNeo commented:
> For sure. Had compared the commits with github. It was identical. I had also tried it back and forth several times. `dev` always worked and `feature/localization` crashed every time.
> 
> I will try to analyze the problem tomorrow.

can you try directly from my forked repository?
https://github.com/DanieleNeo/Daikin-Rotex-HPSU-CAN-Seriell/tree/feature/localization

Before pushing I tried all language, two different board and two environment (development and "production"). Try also to elevate log to debug, let me know.

D.
